### PR TITLE
fixed testString to disallow any extra characters from passing. Also …

### LINKED ISSUE #35557: Add test to disallow extra characters in Applied Accessibility: Standardize Times with the HTML5 datetime Attribute
In the challenge [Applied Accessibility: Standardize Times with the HTML5 datetime Attribute](https://www.freecodecamp.rocks/learn/responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute), if an extra character/characters is/are added in the code between the `</sup>` and `</time>` tags, the challenge still passes. The buggy code looks like this:
```html
<p>Thank you to everyone for responding to Master Camper Cat's survey. The best day to host the vaunted Mortal Kombat tournament is <time datetime="2016-09-15"> Thursday, September 15<sup>th</sup>randomweirdcharacters</time>. May the best ninja win!</p>
```
Note the `randomweirdcharacters` part, we want a test to disallow it.

Refer to the discussion on PR #35553

### PR DESCRIPTION
**Pull request by Sfailla**
…changed markup in testString and text

- [x] I have read [freeCodeCamp's contribution guidelines](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/CONTRIBUTING.md).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `master` branch of freeCodeCamp.
- [x] None of my changes are plagiarized from another source without proper attribution.
- [x] All the files I changed are in the same world language (for example: only English changes, or only Chinese changes, etc.)
- [x] My changes do not use shortened URLs or affiliate links.

Closes #35557

### CODE REVIEW

File: curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md
Comment (RandellDawson) [inline curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: This test has always been a problem for me.  It shows passing without any modifications to the code, because there are other `time` elements with `datetime` attributes already.  I suggest the following change to fix this:
```suggestion
    testString: assert($('time[datetime]').length === 4);
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@Stevo99 Your changes look good.  I do have one suggestion which is not related to your changes, but should improve an existing test.  Just copy/paste my changes and test it out locally to make sure it does not create problems.  If everything tests out OK, you can simply click the Commit suggestion button and it will make the commit for you on GitHub without having to push again.

File: curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md
Comment (Stevo99) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: Ok I made the changes. It works but the only thing i noticed is when the first <code>datetime</code> attr is left empty the challenge still passes for that individual test but not the group of tests. This is due to the other test checking if its equal to the specific date . <code>Your datetime attribute should be set to a value of 2016-09-15.</code>

can I try to fix it so that the individual test won't pass if its empty? or is it good enough?

Comment (Stevo99) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: it works in the sense that it won't pass right away anymore but it will pass if there is an empty <code>datetime</code> attribute

Comment (RandellDawson) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: Which test?

Comment (Stevo99) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: take a look at this code if we do 
<code>assert($('time[datetime]').length === 4 && $('time').attr('datetime') !== "");</code>
the code then works perfectly as it should!  let me know what you think

Comment (Stevo99) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: the test is 2nd one on Applied Accessibility: Standardize Times with the HTML5 datetime Attribute :
Your <code>time</code> tag should have a <code>datetime</code> attribute that is not empty.

**Comment by RandellDawson** [comment]
@Stevo99 Did you mean to close this PR?  I have reopened it for you.

> Hey Randell I noticed all the other tests in the Applied-Accessability have the extra message text in the textString. Can I go thru and remove all the unnecessary markup??

You mean go through all the tests in that section to remove the extra argument?  There is actually an [open issue about doing this](https://github.com/freeCodeCamp/freeCodeCamp/issues/34969) and I have been working on a script which strips them out of applicable challenges.  I say "applicable" because there are some challenges in certain sections (like the back-end challenges) which work differently and still need the message argument.

I think doing this manually would be a pain and prone to creating errors, but you are more than welcome to try.  Just reference that issue number in the PR like `Related to issue #putNumberHere`

**Comment by Stevo99** [comment]
no I didnt mean to close it.  Just new to the process and I think I may have closed it on accident.  Thanks for opening it back up.
I guess Ill just hold off because it seems your way is gonna be way more efficient.  If there is anything else you need feel free to let me know.  Ill search the issues and see if I can find something to work on!

**Comment by RandellDawson** [comment]
> Ill search the issues and see if I can find something to work on!

Great!

**Review by moT01 — APPROVED** [review]
I think I'm satisfied with this - there's still some edge cases I see, but I think this will work - If there's any more troubles in the future, we can deal with them then.

**Comment by Stevo99** [comment]
awesome.. thanks!!

**Comment by RandellDawson** [comment]
> there's still some edge cases I see, but I think this will work 

@moT01 Can you explain some of the edge cases?

**Comment by moT01** [comment]
They really aren't a big deal I don't think - and probably outside the scope of the issue. It's more just how the tests are in general - that they don't narrow it down enough to check for the exact correct things - and the fact that there's multiple `time` elements already on the page that might confuse a user. 

So, for instance, if a user changed the first existing `time` element to the correct text and `datetime`, it would pass two of the tests. If they then added another `time` element with a `datetime` it would pass all four. 

But the instructions are pretty clear, and it says to "add code above/below this line" - It seems like a long shot that someone would pass the tests with the wrong code. So I'm thinking to just leave it like this for now.

**Comment by Stevo99** [comment]
@RandellDawson I made those changes you requested a while ago.  Did I maybe miss something or not do it correctly?  It says you still request changes, but I am confused as to what u need.  Let me know when you can thanks!

**Comment by RandellDawson** [comment]
@Stevo99 Still reviewing this one, but there are many older PRs which have taken priority.  I will respond back when I get some time to read and test your suggestions further.

**Comment by Stevo99** [comment]
ok great!

**Comment by RandellDawson** [comment]
@Stevo99 Just getting back to reviewing this again.  So you made a commit which added `&& $('time').attr('datetime') !== ""` and then you made another commit which removed `&& $('time').attr('datetime') !== ""`.  What was the rationale in removing it?  You previously stated it was needed.

**Comment by Stevo99** [comment]
yea, honestly that was the first open source project I worked on and I was clueless.  Yea I knew that code was correct but It wasn't what you originally told me to change it to.  Then I got worried that you weren't accepting the merge because I added that line.  So I removed it.  I eventually realized that you just needed time to get to it but by then I flip flopped and didnt wanna change anything till I heard back.  If its ok Ill change it back now.

**Comment by RandellDawson** [comment]
@Stevo99 After adding back in the part you took out and doing more testing, I am in agreement with @moT01 about the edge cases and the fact that we really only care that the user changes the one specific `p` element for the purposes of the test.  

I want to see what both of you think about the following changes to the tests.  I added a `tear-down` div to make the test code look cleaner.  

**New teardown div added below the current div with id='html-seed' within the section with id='challenge-seed'**
````html
<div id='html-teardown'>

```html
<script>
const pElement = $("article > p")
  .filter((_, elem) => $(elem).text().includes("Thank you to everyone for responding to Master Camper Cat's survey."));
const timeElement = $(pElement).find("time");
const datetimeAttr = $(timeElement).attr("datetime");
</script>
```

</div>
````

**New tests section**
```yml
tests:
  - text: Your code should have a <code>p</code> element which includes the text "Thank you to everyone for responding to Master Camper Cat's survey." and include a <code>time</code> element.
    testString: assert(timeElement.length);
  - text: Your added <code>time</code> tags should wrap around the text "Thursday, September 15&lt;sup&gt;th&lt;/sup&gt;".
    testString: assert(timeElement.length && $(timeElement).html().trim() === "Thursday, September 15<sup>th</sup>");
  - text: Your added <code>time</code> tag should have a <code>datetime</code> attribute that is not empty.
    testString: assert(datetimeAttr && datetimeAttr.length);
  - text: Your added <code>datetime</code> attribute should be set to a value of 2016-09-15.
    testString: assert(datetimeAttr === "2016-09-15");
  - text: Your added <code>time</code> element should have a closing tag.
    testString: assert(timeElement[0].outerHTML.match(/<\/time>/g));
```

**Comment by moT01** [comment]
Is there an explanation anywhere about what exactly the `html-teardown` and other various sections like this do or how/when they are executed?

**Comment by Stevo99** [comment]
@RandellDawson I just tested it and it works great.  All the test cases respond as they should!!

**Comment by RandellDawson** [comment]
@Stevo99 Then go ahead and make those changes and I will get @moT01 and another mod to review since I made the recommended changes.  Thanks for your patience while we have worked through the review process.

**Comment by Stevo99** [comment]
ok thanks!

**Comment by Stevo99** [comment]
@RandellDawson ok all done. The second commit was to fix a merge issue.

**Comment by RandellDawson** [comment]
@Stevo99 You need to make sure you run `npm run test:curriculum` and make sure all the tests pass before pushing your changes in the future.  The build is failing because you did not correctly resolve the merge conflict.  See below where I have marked which parts should not be in the file.

image

Also, the div with id='html-teardown' was not supposed to be nested inside the div with id='html-seed'.  Instead, it was supposed to be a separate div below the div with id='html-seed'.  

See the [Challenge Template section](https://github.com/freeCodeCamp/freeCodeCamp/blob/master/docs/how-to-work-on-coding-challenges.md#challenge-template) of the How to Work on Coding Challenges documentation to see which shows where to put the teardown div.

If you have further questions or problems you can reach out to the [Contributors Room on Gitter](https://gitter.im/FreeCodeCamp/Contributors)

**Comment by Stevo99** [comment]
yea Im sorry. I was just looking at it now

**Review by RandellDawson — APPROVED** [review]

**Comment by RandellDawson** [comment]
@Stevo99 I cleaned up the git history so there is only one commit now.  I will get two other mods to review the PR and hopefully we can get this merged soon.

**Comment by Stevo99** [comment]
@RandellDawson  Ok I appreciate that!  Thanks for all your help!

File: curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md
Comment (ojeytonwilliams) [inline curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]:
```suggestion
    testString: assert(/<\/h2>.*<\/time>.*<section>/s.test(code));
```
jQuery doesn't work here, since the browser will automatically close the opening `<time>` tag.  If you have an opening tag, `outerHTML` will always match.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
Only the one problem, looks good otherwise.  Thanks for your patience, @Stevo99, we're nearly there!

File: curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md
Comment (Stevo99) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: Ok i will make changes now

Comment (ojeytonwilliams) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: You can always click 'Commit suggestion' if you're happy with the change

Comment (Stevo99) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: ok yea I was wondering if I could do that lol

Comment (ojeytonwilliams) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: Yeah, weirdly you can't reliably commit suggestions on someone else's branch (even when you have write access and can push commits), but the author always can.  Saves loads of time.

**Review by ojeytonwilliams — APPROVED** [review]

**Comment by ojeytonwilliams** [comment]
Thanks, @Stevo99, it LGTM!  We're just going to wait for one final review, but everything seems fine to me.

File: curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md
Comment (RandellDawson) [inline curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: We do not even need this test.
```suggestion
```

File: curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md
Comment (RandellDawson) [inline curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: We do not even need this test.
```suggestion
```

**Review by RandellDawson — CHANGES_REQUESTED** [review]
@St3ve89 @ojeytonwilliams  I originally hesitated about even keeping the last test, which is why I used the `outerHTML` reference to begin with.  Why?  Because the 2nd test will fail anyway if the user does not add the `</time>` tag.  I suggest we just git rid of the last test.

File: curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md
Comment (Stevo99) [reply curriculum/challenges/english/01-responsive-web-design/applied-accessibility/standardize-times-with-the-html5-datetime-attribute.english.md]: ok just changed it!

**Comment by RandellDawson** [comment]
@Stevo99 You must delete both lines or the Travis build tests will fail (as they are now).

**Comment by ojeytonwilliams** [comment]
@RandellDawson I wasn't happy that it would pass while there was no tag, but removing it entirely does solve that problem :smile:

**Comment by Stevo99** [comment]
Ok sorry I was thinking it was weird not removing the message as well.  I just figured you knew better than me and it was there for a reason.  Next time Ill ask if I have a gut feeling to keep avoiding simple mistakes.

**Comment by RandellDawson** [comment]
@Stevo99 I marked both lines to be removed.  You only committed one of the requested changes.  That is all.

**Comment by Stevo99** [comment]
Ok got it

**Comment by RandellDawson** [comment]
@Stevo99 OK, now we just need one more moderator to review and approve.

**Comment by Stevo99** [comment]
ok sounds good!  thanks

**Review by RandellDawson — APPROVED** [review]

**Review by Manish-Giri — APPROVED** [review]
Tested locally, and tried a few edge cases, looks like this covers them all.

@Stevo99 @ojeytonwilliams @RandellDawson  Good work!

**Pull request merged**
